src/KubeOps.Operator/Reconciliation/Reconciler.cs

```diff
@@ -0,0 +1,202 @@
+                        token: cancellationToken);
+                }
+
+                return await ReconcileEntity(reconciliationContext.Entity, cancellationToken);
+            case { Metadata: { DeletionTimestamp: not null, Finalizers.Count: > 0 } }:
+                return await ReconcileFinalizersSequential(reconciliationContext.Entity, cancellationToken);
+            default:
+                return ReconciliationResult<TEntity>.Success(reconciliationContext.Entity);
+        }
+    }
+
+    private async Task<ReconciliationResult<TEntity>> ReconcileDeletion(ReconciliationContext<TEntity> reconciliationContext, CancellationToken cancellationToken)
+    {
+        await entityQueue
+            .Remove(
+                reconciliationContext.Entity,
+                cancellationToken);
+
+        await using var scope = serviceProvider.CreateAsyncScope();
+        var controller = scope.ServiceProvider.GetRequiredService<IEntityController<TEntity>>();
+        var result = await controller.DeletedAsync(reconciliationContext.Entity, cancellationToken);
+
+        if (result.IsSuccess)
+        {
+            await _entityCache.RemoveAsync(reconciliationContext.Entity.Uid(), token: cancellationToken);
+        }
+
+        return result;
+    }
+
+    private async Task<ReconciliationResult<TEntity>> ReconcileEntity(TEntity entity, CancellationToken cancellationToken)
+    {
+        await entityQueue
+            .Remove(
+                entity,
+                cancellationToken);
+
+        await using var scope = serviceProvider.CreateAsyncScope();
+
+        if (operatorSettings.AutoAttachFinalizers)
+        {
+            var finalizers = scope.ServiceProvider.GetKeyedServices<IEntityFinalizer<TEntity>>(KeyedService.AnyKey);
+
+            foreach (var finalizer in finalizers)
+            {
+                entity.AddFinalizer(finalizer.GetIdentifierName(entity));
+            }
+
+            entity = await client.UpdateAsync(entity, cancellationToken);
+        }
+
+        var controller = scope.ServiceProvider.GetRequiredService<IEntityController<TEntity>>();
+        return await controller.ReconcileAsync(entity, cancellationToken);
+    }
+
+    private async Task<ReconciliationResult<TEntity>> ReconcileFinalizersSequential(TEntity entity, CancellationToken cancellationToken)
+    {
+        await entityQueue
+            .Remove(
+                entity,
+                cancellationToken);
+
+        await using var scope = serviceProvider.CreateAsyncScope();
+
+        // the condition to call ReconcileFinalizersSequentialAsync is:
+        // { Metadata: { DeletionTimestamp: not null, Finalizers.Count: > 0 } }
+        // which implies that there is at least a single finalizer
+        var identifier = entity.Finalizers()[0];
+
+        if (scope.ServiceProvider.GetKeyedService<IEntityFinalizer<TEntity>>(identifier) is not
+            { } finalizer)
+        {
+            logger.LogDebug(
```

Comment:
I personally would place this as a "warning" or at least information because this could indicate a split-brain problem when your own operator is not able to finalize your own custom objects. There are multiple use-cases but yes.